core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```diff
@@ -0,0 +1,192 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.percolator;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.*;
+import org.apache.lucene.queries.TermsQuery;
+import org.apache.lucene.search.*;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.mapper.ParseContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Utility to extract query terms from queries and create queries from documents.
+ */
+public class QueryMetadataService {
+
+    public static String QUERY_METADATA = "_query_metadata_";
+    public static String QUERY_METADATA_FIELD_PREFIX = QUERY_METADATA + "field_";
+    public static String QUERY_METADATA_FIELD_UNKNOWN = QUERY_METADATA + "unknown";
+    public static FieldType QUERY_METADATA_FIELD_TYPE = new FieldType();
+
+    static {
+        QUERY_METADATA_FIELD_TYPE.setTokenized(false);
+        QUERY_METADATA_FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+        QUERY_METADATA_FIELD_TYPE.freeze();
+    }
+
+    /**
+     * Extracts all terms from the specified query and adds it to the specified document.
+     */
+    public void extractQueryMetadata(Query query, ParseContext.Document document) {
+        List<Term> queryTerms;
+        try {
+            queryTerms = extractQueryMetadata(query);
+        } catch (IllegalArgumentException e) {
+            document.add(new Field(QUERY_METADATA_FIELD_UNKNOWN, new BytesRef(), QUERY_METADATA_FIELD_TYPE));
+            return;
+        }
+        for (Term term : queryTerms) {
+            document.add(new Field(QUERY_METADATA_FIELD_PREFIX + term.field(), term.bytes(), QUERY_METADATA_FIELD_TYPE));
+        }
+    }
+
+    /**
+     * Extracts all query terms from the provided query and adds it to specified list.
+     *
+     * From boolean query with no should clauses or phrase queries only the the longest term are selected,
+     * since that those terms are likely to be the rarest. Boolean query's must_not clauses are always ignored ignored.
+     *
+     * If from part of the query, no query terms can be extracted then term extraction is stopped and
+     * an IllegalArgumentException is thrown.
```

Comment:
maybe we need a more specific exception? I'm afraid that if eg. one of the methods that extractQueryMetadata calls throws an IAE then term extraction will silently fail